### PR TITLE
test: make tests work in python 3.12+

### PR DESCRIPTION
**Pull request by lengau**
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?

-----

### CODE REVIEW

**Review by syu-w — APPROVED** [review]

File: tests/unit/plugins/test_base.py [hunk 1]
```diff
@@ -78,13 +78,18 @@ class FaultyPlugin(Plugin):
...
 
-    with pytest.raises(TypeError) as raised:
-        FaultyPlugin(properties=None, part_info=part_info)  # type: ignore[reportGeneralTypeIssues]
-    assert str(raised.value) == (
-        "Can't instantiate abstract class FaultyPlugin with abstract methods "
-        "get_build_commands, get_build_environment, get_build_packages, get_build_snaps"
+    expected = (
+        r"^Can't instantiate abstract class FaultyPlugin with(out an implementation "
+        r"for)? abstract methods '?get_build_commands'?, '?get_build_environment'?, "
+        r"'?get_build_packages'?, '?get_build_snaps'?$"
```
Comment (syu-w) [inline tests/unit/plugins/test_base.py:84]: Not sure in this case we want to try match all of them

File: tests/unit/plugins/test_base.py [hunk 2]
```diff
@@ -78,13 +78,18 @@ class FaultyPlugin(Plugin):
...
     )
 
+    with pytest.raises(
+        TypeError,
+        match=expected,
+    ):
```
Comment (cmatsuoka) [inline tests/unit/plugins/test_base.py:90]: Can we have that in a single line?

**Review by cmatsuoka — COMMENTED** [review]
Thanks! (but it seems that we have a linter error)

File: tests/unit/plugins/test_base.py (hunk 2 above)
Comment (lengau) [reply tests/unit/plugins/test_base.py]: Done!

**Review by cmatsuoka — APPROVED** [review]
Thanks!

**Pull request merged**